api/layer/multipart_upload.go

```diff
@@ -0,0 +1,494 @@
+package layer
+
+import (
+	"context"
+	"io"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/nspcc-dev/neofs-api-go/pkg/object"
+	"github.com/nspcc-dev/neofs-api-go/pkg/owner"
+	"github.com/nspcc-dev/neofs-s3-gw/api/data"
+	"github.com/nspcc-dev/neofs-s3-gw/api/errors"
+	"go.uber.org/zap"
+)
+
+const (
+	UploadIDAttributeName         = "S3-Upload-Id"
+	UploadPartNumberAttributeName = "S3-Upload-Part-Number"
+	UploadKeyAttributeName        = "S3-Upload-Key"
+	UploadHasTags                 = "S3-Upload-Has-Tags"
+	UploadHasACL                  = "S3-Upload-Has-ACL"
+	UploadPartKeyPrefix           = ".upload-"
+
+	uploadMinSize = 5 * 1048576       // 5MB
+	uploadMaxSize = 5 * 1099511627776 // 5TB
```

Comment:
5TB is max object size not max part size https://docs.aws.amazon.com/AmazonS3/latest/userguide/qfacts.html